src/vms/pvm/etna-builder/spend-reducers/handleFeeAndChange.ts

```diff
@@ -0,0 +1,101 @@
+import {
+  BigIntPr,
+  Id,
+  OutputOwners,
+  TransferOutput,
+  TransferableOutput,
+} from '../../../../serializable';
+import type { Context } from '../../../context';
+import type { SpendReducerFunction } from './types';
+
+/**
+ * Determines if the fee can be covered by the excess AVAX.
+ *
+ * @returns {boolean} - Whether the excess AVAX exceeds the fee. `true` greater than the fee, `false` if equal.
+ * @throws {Error} - If the excess AVAX is less than the required fee.
+ */
+const canPayFeeAndNeedsChange = (
+  excessAVAX: bigint,
+  requiredFee: bigint,
+  context: Context,
+): boolean => {
+  // Not enough funds to pay the fee.
+  if (excessAVAX < requiredFee) {
+    throw new Error(
+      `Insufficient funds: provided UTXOs need ${
+        requiredFee - excessAVAX
+      } more nAVAX (asset id: ${context.avaxAssetID})`,
+    );
+  }
+
+  // No need to add a change to change output.
+  // Just burn the fee.
+  if (excessAVAX === requiredFee) {
+    return false;
+  }
+
+  return true;
+};
+
+export const handleFeeAndChange: SpendReducerFunction = (
+  state,
+  spendHelper,
+  context,
+) => {
+  // Use the change owner override if it exists, otherwise use the default change owner.
+  // This is used on "import" transactions.
+  const changeOwners =
+    state.ownerOverride ??
+    OutputOwners.fromNative(state.spendOptions.changeAddresses);
+
+  const requiredFee = spendHelper.calculateFee();
+
+  // Checks for an existing change output that is for the AVAX asset assigned to the change owner.
+  const hasExistingChangeOutput: boolean = spendHelper.hasChangeOutput(
+    context.avaxAssetID,
+    changeOwners,
+  );
+
+  if (canPayFeeAndNeedsChange(state.excessAVAX, requiredFee, context)) {
+    if (hasExistingChangeOutput) {
+      // Excess exceeds fee, return the change.
+      // This output will get consolidated with the existing output.
+      spendHelper.addChangeOutput(
+        new TransferableOutput(
+          Id.fromString(context.avaxAssetID),
+          new TransferOutput(
+            new BigIntPr(state.excessAVAX - requiredFee),
+            changeOwners,
+          ),
+        ),
+      );
+    } else {
+      // Calculate the fee with a temporary output complexity
+      // as if we added the change output.
+      const requiredFeeWithChangeOutput =
+        spendHelper.calculateFeeWithTemporaryOutputComplexity(
+          new TransferableOutput(
+            Id.fromString(context.avaxAssetID),
+            new TransferOutput(new BigIntPr(0n), changeOwners),
+          ),
+        );
+
+      // If the excess AVAX is greater than the new fee, add a change output.
+      // Otherwise, ignore and burn the excess because it can't be returned
+      // (ie we can't pay the fee to return the excess).
+      if (state.excessAVAX > requiredFeeWithChangeOutput) {
```

Comment:
I don't fully understand why we can exit in the other cases. Can you clarify in a comment?

Reply:
Basically at this stage, we have determined that the user should get change back, but if we determine that adding an output for the change increases the fee to the point that you can't pay for your own change back, there is no point in returning the change, so the excess just gets burnt.